<a href="https://colab.research.google.com/github/FareedaAbdelaal/ML-55-9988/blob/main/ML_Task_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from google.colab import files
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_absolute_error, mean_squared_error


In [13]:
# Upload and Load Dataset

uploaded = files.upload()

# Load dataset
df = pd.read_csv("Advertising.csv")

print("===== First 5 Rows of Dataset =====")
print(df.head())


Saving Advertising.csv to Advertising.csv
===== First 5 Rows of Dataset =====
      TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9


In [14]:
# Data Preparation


# Features (X) and Target (y)
X = df.drop("sales", axis=1)
y = df["sales"]

print("\n===== Features =====")
print(X.head())

print("\n===== Target =====")
print(y.head())


# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.3,
    random_state=101
)

print("\n===== Train/Test Split =====")
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)




===== Features =====
      TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

===== Target =====
0    22.1
1    10.4
2     9.3
3    18.5
4    12.9
Name: sales, dtype: float64

===== Train/Test Split =====
X_train shape: (140, 3)
X_test shape: (60, 3)
y_train shape: (140,)
y_test shape: (60,)


In [15]:
# Linear Regression Model

# Create Model
linear_model = LinearRegression()

# Train Model
linear_model.fit(X_train, y_train)

# Predictions
linear_predictions = linear_model.predict(X_test)

# Evaluation Metrics
linear_MAE = mean_absolute_error(y_test, linear_predictions)
linear_MSE = mean_squared_error(y_test, linear_predictions)
linear_RMSE = np.sqrt(linear_MSE)

print("\n===== Linear Regression Results =====")
print("MAE:", linear_MAE)
print("MSE:", linear_MSE)
print("RMSE:", linear_RMSE)



===== Linear Regression Results =====
MAE: 1.213745773614481
MSE: 2.2987166978863796
RMSE: 1.5161519375993884


In [16]:
#Polynomial Regression Model

# Create Polynomial Features (degree = 2)
poly_converter = PolynomialFeatures(
    degree=2,
    include_bias=False
)

# Fit ONLY on training data (avoids data leakage)
poly_converter.fit(X_train)

# Transform train and test data separately
X_train_poly = poly_converter.transform(X_train)
X_test_poly = poly_converter.transform(X_test)

# Create Model
poly_model = LinearRegression()

# Train Model
poly_model.fit(X_train_poly, y_train)

# Predictions
poly_predictions = poly_model.predict(X_test_poly)

# Evaluation Metrics
poly_MAE = mean_absolute_error(y_test, poly_predictions)
poly_MSE = mean_squared_error(y_test, poly_predictions)
poly_RMSE = np.sqrt(poly_MSE)

print("\n===== Polynomial Regression Results =====")
print("MAE:", poly_MAE)
print("MSE:", poly_MSE)
print("RMSE:", poly_RMSE)



===== Polynomial Regression Results =====
MAE: 0.4896798044803838
MSE: 0.4417505510403753
RMSE: 0.6646431757269274


In [17]:
# =========================================
# 6. Model Comparison
# =========================================

comparison = pd.DataFrame({
    "Model": ["Linear Regression", "Polynomial Regression"],
    "MAE": [linear_MAE, poly_MAE],
    "MSE": [linear_MSE, poly_MSE],
    "RMSE": [linear_RMSE, poly_RMSE]
})

print("\n===== Model Comparison =====")
print(comparison)


# Identify Better Model
if poly_RMSE < linear_RMSE:
    print("\nPolynomial Regression performs better because it has lower RMSE.")
else:
    print("\nLinear Regression performs better because it has lower RMSE.")



===== Model Comparison =====
                   Model       MAE       MSE      RMSE
0      Linear Regression  1.213746  2.298717  1.516152
1  Polynomial Regression  0.489680  0.441751  0.664643

Polynomial Regression performs better because it has lower RMSE.


In [18]:
# Coefficients Interpretation


print("\n===== Linear Regression Coefficients =====")

coef_df = pd.DataFrame({
    "Feature": X.columns,
    "Coefficient": linear_model.coef_
})

print(coef_df)

print("\nIntercept:", linear_model.intercept_)


# Simple Interpretation
print("\n===== Interpretation =====")
print("Positive coefficient → increasing this feature increases sales.")
print("Negative coefficient → increasing this feature decreases sales.")
print("Larger absolute value → stronger effect on prediction.")


===== Linear Regression Coefficients =====
     Feature  Coefficient
0         TV     0.044696
1      radio     0.187566
2  newspaper    -0.000323

Intercept: 3.1515267680706547

===== Interpretation =====
Positive coefficient → increasing this feature increases sales.
Negative coefficient → increasing this feature decreases sales.
Larger absolute value → stronger effect on prediction.


The results show that Polynomial Regression performed better than Linear Regression on the Advertising dataset because it achieved lower error values in MAE, MSE, and RMSE. Linear Regression had an RMSE of 1.516, while Polynomial Regression reduced it significantly to 0.664, indicating more accurate predictions. This suggests that the relationship between advertising features and sales is not perfectly linear, and Polynomial Regression captures the pattern more effectively.